snuba/query/expressions.py

```diff
@@ -0,0 +1,123 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Callable, Iterable, Optional, Sequence
+
+from snuba.query.nodes import AliasedNode, Node
+
+
+class Expression(AliasedNode, ABC):
+    """
+    Abstract representation of a Query node that can be evaluated as a column.
+    This can be a simple column or a nested expression, but not a condition.
+    """
+    @abstractmethod
+    def map(self, closure: Callable[[Expression], Expression]) -> Expression:
```

Comment:
I tried a few ways to put map and filter in a generic Node interface but I don't think there was a reasonable solution to make the type checker enforce that the return value of the closure would always be a safe replacement.
Example:
1) if these were defined without generics at Node level, the closure would return a Node, which means there would be no way to make the type checker ensure that a closure applied to an Expression would not return a Condition
2) I tried to have a SubTree class that was generic with regards to the Node type, but the result is that Expression would have to materialize that parameter with itself: class Expresison[SubTree[Expression]] (which is impossible).

The main issue is that this is a tree structure, thus every edit operation that traverses the tree need to return something of the same type of self.